src/mono/mono/mini/method-to-ir.c

```diff
@@ -9546,6 +9558,63 @@ mono_method_to_ir (MonoCompile *cfg, MonoMethod *method, MonoBasicBlock *start_b
 				}
 			}
 
+			// ASSUME
+			//   interface ISomeIface { void Method(); }
+			//   struct SomeStruct : ISomeIface {...}
+			// OPTIMIZE
+			//   box SomeStruct
+			//   callvirt instance ISomeIface::Method()
+			// TO
+			//   call SomeStruct::Method()
+			guchar* callvirt_ip;
+			guint32 callvirt_proc_token;
+			if((callvirt_ip = il_read_callvirt(next_ip, end, &callvirt_proc_token)) && 
+				ip_in_bb(cfg, cfg->cbb, callvirt_ip) ) 
+			{
+				MonoMethod* iface_method = mini_get_method(cfg, method, callvirt_proc_token, NULL, generic_context);
+				MonoMethodSignature* iface_method_sig = mono_method_signature_internal(iface_method);
+
+				if(	val &&
+					val->flags != MONO_INST_FAULT && // not null
+					iface_method_sig != NULL && // callee signture is healthy
+					iface_method_sig->hasthis && 
+					iface_method_sig->param_count == 0) // the callee has no args (other than this)
+				{
+					if (!m_class_is_inited (iface_method->klass)) {
+						if (!mono_class_init_internal (iface_method->klass))
+							TYPE_LOAD_ERROR (iface_method->klass);
+					}
+
+					MonoMethod* struct_method = mono_class_get_virtual_method(klass, iface_method, error);
+					
+					if(is_ok(error)) {
+						MonoClass* struct_obj = mono_method_get_class(struct_method);
+
+						if(	!struct_method ||
+							!mono_method_can_access_method(method, struct_method) ||
+							mono_class_is_abstract(struct_obj)) {
+							// do not optimize, let full callvirt deal with it
+						} else if(val->flags & MONO_INST_INDIRECT) {
+							*sp++ = val;
+							cmethod_override = struct_method;
+							break;
+						} else {
+							MonoInst* srcvar = get_vreg_to_inst(cfg, val->dreg);
+							if (!srcvar)
+								srcvar = mono_compile_create_var_for_vreg (cfg, m_class_get_byval_arg (klass), OP_LOCAL, val->dreg);
+
+							EMIT_NEW_VARLOADA (cfg, ins, srcvar, m_class_get_byval_arg(klass));
+							*sp++= ins;
+							cmethod_override = struct_method;
+							break;
+						}
+						
+					} else {
+						mono_error_cleanup(error);
```

Comment:
MonoError is meant to be one-shot - once it's set and cleaned up it's marked so that it can't be set again.  (We may need to revisit this design some day but that's how it is right now)

In the case of recoverable errors, it's better to use a local:

```c
ERROR_DECL(local_error);
result = some_call (args..., local_error);
if (!is_ok (local_error)) {
   mono_error_cleanup (local_error);
   /* do something else */
} else {
   /* do something good */
}
```

Reply:
Thanks for pointing this one out. I will switch to the local handler.